src/contracts/core/DelegationManager.sol

```diff
@@ -528,131 +633,224 @@ contract DelegationManager is
         );
-
         require(
             msg.sender == withdrawal.withdrawer,
             "DelegationManager._completeQueuedWithdrawal: only withdrawer can complete action"
         );
+        require(
+            tokens.length == withdrawal.strategies.length,
+            "DelegationManager._completeQueuedWithdrawal: input length mismatch"
+        );
+
+        // read delegated operator's totalMagnitudes at time of withdrawal to scale shares again if any slashing has occurred
+        // during withdrawal delay period
+        uint64[] memory totalMagnitudes = allocationManager.getTotalMagnitudesAtTimestamp({
+            operator: withdrawal.delegatedTo,
+            strategies: withdrawal.strategies,
+            timestamp: withdrawal.startTimestamp + SlashingConstants.DEALLOCATION_DELAY
+        });
 
         if (receiveAsTokens) {
-            require(
-                tokens.length == withdrawal.strategies.length,
-                "DelegationManager._completeQueuedWithdrawal: input length mismatch"
-            );
+            // complete the withdrawal by converting shares to tokens
+            _completeReceiveAsTokens(withdrawal, tokens, totalMagnitudes);
+        } else {
+            // Award shares back in StrategyManager/EigenPodManager.
+            _completeReceiveAsShares(withdrawal, tokens, totalMagnitudes);
         }
 
         // Remove `withdrawalRoot` from pending roots
         delete pendingWithdrawals[withdrawalRoot];
+        emit WithdrawalCompleted(withdrawalRoot);
+    }
 
-        if (receiveAsTokens) {
-            // Finalize action by converting shares to tokens for each strategy, or
-            // by re-awarding shares in each strategy.
-            for (uint256 i = 0; i < withdrawal.strategies.length;) {
-                require(
-                    withdrawal.startBlock + strategyWithdrawalDelayBlocks[withdrawal.strategies[i]] <= block.number,
-                    "DelegationManager._completeQueuedWithdrawal: withdrawalDelayBlocks period has not yet passed for this strategy"
-                );
+    /**
+     * @dev This function completes a queued withdrawal for a staker by converting shares to tokens and transferring to the withdrawer.
+     * This will apply any slashing that has occurred since the scaledShares were initially set in the Withdrawal struct
+     * by reading the original delegated operator's totalMagnitude at the time of withdrawal completion.
+     */
+    function _completeReceiveAsTokens(
+        Withdrawal calldata withdrawal,
+        IERC20[] calldata tokens,
+        uint64[] memory totalMagnitudes
+    ) internal {
+        // Finalize action by converting scaled shares to tokens for each strategy, or
+        // by re-awarding shares in each strategy.
+        for (uint256 i = 0; i < withdrawal.strategies.length;) {
+            require(
+                withdrawal.startTimestamp + strategyWithdrawalDelays[withdrawal.strategies[i]] <= block.timestamp,
```

Comment:
we need to change this

Reply:
removed and just using the `MIN_WITHDRAWAL_DELAY`